pkg/node/manager/manager.go

```diff
@@ -332,134 +333,138 @@ func (m *manager) legacyNodeIpBehavior() bool {
-		// encapsulation. In encryption case we also want to use vxlan device
-		// to create symmetric traffic when sending nodeIP->pod and pod->nodeIP.
-		if address.Type == addressing.NodeCiliumInternalIP || m.conf.EncryptionEnabled() ||
-			option.Config.EnableHostFirewall || option.Config.JoinCluster {
-			tunnelIP = nodeIP
-		}
-
 		if option.Config.NodeIpsetNeeded() && address.Type == addressing.NodeInternalIP {
 			iptables.AddToNodeIpset(address.IP)
 		}
 
-		if skipIPCache(address) {
+		if m.nodeAddressSkipsIPCache(address) {
 			continue
 		}
 
-		// If we are doing encryption, but not node based encryption, then do not
-		// add a key to the nodeIPs so that we avoid a trip through stack and attempting
-		// to encrypt something we know does not have an encryption policy installed
-		// in the datapath. By setting key=0 and tunnelIP this will result in traffic
-		// being sent unencrypted over overlay device.
-		if (!m.conf.NodeEncryptionEnabled() &&
-			(address.Type == addressing.NodeExternalIP || address.Type == addressing.NodeInternalIP)) ||
-			// Also ignore any remote node's key if the local node opted to not perform
-			// node-to-node encryption
-			node.GetOptOutNodeEncryption() {
-			key = 0
+		var tunnelIP netip.Addr
+		if m.nodeAddressHasTunnelIP(address) {
+			tunnelIP = nodeIP
 		}
 
-		var prefix netip.Prefix
-		if v4 := address.IP.To4(); v4 != nil {
-			prefix = ip.IPToNetPrefix(v4)
-		} else {
-			prefix = ip.IPToNetPrefix(address.IP.To16())
-		}
-		ipAddrStr := prefix.String()
-		_, err := m.ipcache.Upsert(ipAddrStr, tunnelIP, key, nil, ipcache.Identity{
-			ID:     remoteHostIdentity,
-			Source: n.Source,
-		})
-		resource := ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindNode, "", n.Name)
-		m.upsertIntoIDMD(prefix, remoteHostIdentity, resource)
-
-		// Upsert() will return true if the ipcache entry is owned by
-		// the source of the node update that triggered this node
-		// update (kvstore, k8s, ...) The datapath is only updated if
-		// that source of truth is updated.
-		// The only exception are kube-apiserver entries. In that case,
-		// we still want to inform subscribers about changes in auxiliary
-		// data such as for example the health endpoint.
-		overwriteErr := &ipcache.ErrOverwrite{
-			ExistingSrc: source.KubeAPIServer,
-			NewSrc:      n.Source,
+		var key uint8
+		if m.nodeAddressHasEncryptKey(address) {
+			key = n.EncryptionKey
 		}
-		if err != nil && !errors.Is(err, overwriteErr) {
+
+		// We expect the node manager to have a source of either Kubernetes,
+		// CustomResource, or KVStore. Prioritize the KVStore source over the
+		// rest as it is the strongest source.
+		prefix := ip.IPToNetPrefix(address.IP)
+		if !source.AllowOverwrite(m.ipcache.GetMetadataByPrefix(prefix).Source(), n.Source) {
 			dpUpdate = false
-		} else {
-			ipsAdded = append(ipsAdded, ipAddrStr)
+			continue
+		}
+		m.ipcache.UpsertMetadata(prefix, n.Source, resource,
```

Comment:
`dpUpdate=false` here makes sense.

I wonder if we should be Upserting the metadata anyway in this case though, this will quietly hide the updates from the lower priority source.

I don't know what the ordering guarantees are here, but let's say we learn about the same node from k8s and from kvstore. Depending on whether we handle the event for k8s then kvstore, or kvstore then k8s, the ipcache will either end up with two entries for the same node or one entry for the same node. Seems a bit inconsistent?

But then again, I can't think of a concrete example of how this could occur and then lead to some weird follow-on state, it's just that for instance if we could query all internal ipcache state then we might not necessarily see one of the pieces of information used to make the decision about the node events. If we did happen to have a bug in the logic, it could plausibly lead to some wrong behaviour but that's a bit hard to speculate about.

The only other case I could think of is if we ended up supporting scale up + scale down and dynamically enabling+disbaling kvstore. We have no such plans today.

:shrug:

Reply:
(Even if we upserted the metadata from both cases, the nodeIdentityOverride below should only occur iff `dpupdate==true`.)

Reply:
> Depending on whether we handle the event for k8s then kvstore, or kvstore then k8s, the ipcache will either end up with two entries for the same node or one entry for the same node.

Actually this is not quite correct, I was assuming unique `ResourceID`s for the resource sourced from etcd vs. the one sourced from k8s. That's how I think it _should_ work, so we can trace the true source of the data for the ipcache entry back to where it came from. But in practice the ResourceID is set directly earlier in this function for both sources, so there's no way to tell where it came from. It also means that the ipcache entries would be updated / overwritten by the update from the other source.

We also discussed in person previously that we assume that both sources are going to end up propagating the exact same information anyway, so this whole line of consideration is probably more theoretical than practical. I think it's probably fine to chalk this up as "feels a little odd but probably not going to have any consequences" and let my concerns lay to rest :sweat_smile:

Reply:
As a followup, I've [wrote down](https://github.com/cilium/cilium/issues/21142) to come back and rework the resource IDs here so that we can capture kvstore vs k8s.